### ISSUE TITLE
Holding more than one threadsafe function causes SIGABRT on Drop

### ISSUE THREAD

**Issue by pimeys** [napi-rs/napi-rs#518]
An example repo: https://github.com/pimeys/napi-bug-repo

Creating two classes, both holding a threadsafe callback function, will dump core and cause SIGABRT on exit.

```sh
❯ uname -a
Linux naunau 5.11.8 #1-NixOS SMP Sat Mar 20 09:51:14 UTC 2021 x86_64 GNU/Linux
```

```toml
[[package]]
name = "napi"
version = "1.3.3"
```

**Comment by pimeys** [comment]
From the repo, if only creating one instances as in:

```javascript
const api = require(`./libnapi_test.so.node`)

async function main() {
    const a = new api.A((s) => console.log(s));
}
main()
```

The program will exit correctly without aborting.

**Comment by pimeys** [comment]
I compiled nodejs from master with debug symbols to get the stack trace from the core dump:

```gdb
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/nix/store/90illc73xfs933d06daq6d41njs8yh66-glibc-2.32-37/lib/libthread_db.so.1".
Core was generated by `../node/node_g index.js'.
Program terminated with signal SIGABRT, Aborted.
#0  0x00007ff4a7b6c33a in raise () from /nix/store/90illc73xfs933d06daq6d41njs8yh66-glibc-2.32-37/lib/libc.so.6
[Current thread is 1 (Thread 0x7ff4a7b2d7c0 (LWP 23234))]
(gdb) bt
#0  0x00007ff4a7b6c33a in raise () from /nix/store/90illc73xfs933d06daq6d41njs8yh66-glibc-2.32-37/lib/libc.so.6
#1  0x00007ff4a7b56523 in abort () from /nix/store/90illc73xfs933d06daq6d41njs8yh66-glibc-2.32-37/lib/libc.so.6
#2  0x00000000022353e9 in uv_mutex_lock (mutex=0x6ec1e48) at ../deps/uv/src/unix/thread.c:332
#3  0x0000000000fed791 in node::LibuvMutexTraits::mutex_lock (mutex=0x6ec1e48) at ../src/node_mutex.h:164
#4  0x0000000000feea12 in node::MutexBase<node::LibuvMutexTraits>::ScopedLock::ScopedLock (this=0x7ffe68204490, mutex=...)
    at ../src/node_mutex.h:220
#5  0x00000000010cf9f4 in v8impl::(anonymous namespace)::ThreadSafeFunction::Release (this=0x6ec1e20, mode=napi_tsfn_release)
    at ../src/node_api.cc:200
#6  0x00000000010d3324 in napi_release_threadsafe_function (func=0x6ec1e20, mode=napi_tsfn_release) at ../src/node_api.cc:1271
#7  0x00007ff4a52bdae2 in napi::threadsafe_function::{{impl}}::drop<alloc::string::String,napi::threadsafe_function::ErrorStrategy::CalleeHandled> (self=0x6ebcac8) at /home/pimeys/code/napi-rs/napi/src/threadsafe_function.rs:311
#8  0x00007ff4a52bdfa2 in core::ptr::drop_in_place<napi::threadsafe_function::ThreadsafeFunction<alloc::string::String, napi::threadsafe_function::ErrorStrategy::CalleeHandled>> () at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ptr/mod.rs:179
#9  0x00007ff4a52bdc0e in core::ptr::drop_in_place<napi_test::A> ()
    at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ptr/mod.rs:179
#10 0x00007ff4a52bdcb3 in core::ptr::drop_in_place<core::option::Option<napi_test::A>> ()
    at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ptr/mod.rs:179
#11 0x00007ff4a52bdf05 in core::ptr::drop_in_place<napi::js_values::tagged_object::TaggedObject<napi_test::A>> ()
    at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ptr/mod.rs:179
#12 0x00007ff4a52bdb48 in core::ptr::drop_in_place<alloc::boxed::Box<napi::js_values::tagged_object::TaggedObject<napi_test::A>, alloc::alloc::Global>> () at /rustc/2fd73fabe469357a12c2c974c140f67e7cdd76d0/library/core/src/ptr/mod.rs:179
#13 0x00007ff4a52c05ec in napi::env::raw_finalize<napi_test::A> (env=0x6ec0010, finalize_data=0x6ebcac0, finalize_hint=0x0)
    at /home/pimeys/code/napi-rs/napi/src/env.rs:1289
#14 0x00000000010d5a81 in napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}::operator()(napi_env__*) const (__closure=0x7ffe68204710, env=0x6ec0010) at ../src/js_native_api_v8.h:107
#15 0x00000000010d6601 in napi_env__::CallIntoModule<napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}, void (napi_env__*, v8::Local<v8::Value>)>(napi_env__::CallFinalizer(void (*)(napi_env__*, void*, void*), void*, void*)::{lambda(napi_env__*)#1}&&, void (&&)(napi_env__*, v8::Local<v8::Value>)) (this=0x6ec0010, call=..., handle_exception=
    @0x10b3384: {void (napi_env__ *, v8::Local<v8::Value>)} 0x10b3384 <napi_env__::HandleThrow(napi_env__*, v8::Local<v8::Value>)>)
    at ../src/js_native_api_v8.h:95
#16 0x00000000010d5af4 in napi_env__::CallFinalizer (this=0x6ec0010, cb=0x7ff4a52c05b0 <napi::env::raw_finalize<napi_test::A>>,
    data=0x6ebcac0, hint=0x0) at ../src/js_native_api_v8.h:106
#17 0x00000000010a4c13 in v8impl::(anonymous namespace)::RefBase::Finalize (this=0x6e7a0a0, is_env_teardown=true)
    at ../src/js_native_api_v8.cc:298
#18 0x00000000010a4e41 in v8impl::(anonymous namespace)::Reference::Finalize (this=0x6e7a0a0, is_env_teardown=true)
    at ../src/js_native_api_v8.cc:384
#19 0x00000000010d583b in v8impl::RefTracker::FinalizeAll (list=0x6ec0048) at ../src/js_native_api_v8.h:43
#20 0x00000000010d596e in napi_env__::~napi_env__ (this=0x6ec0010, __in_chrg=<optimized out>) at ../src/js_native_api_v8.h:66
#21 0x00000000010d97a0 in node_napi_env__::~node_napi_env__ (this=0x6ec0010, __in_chrg=<optimized out>) at ../src/node_api.cc:17
#22 0x00000000010d97bc in node_napi_env__::~node_napi_env__ (this=0x6ec0010, __in_chrg=<optimized out>) at ../src/node_api.cc:17
#23 0x00000000010b3381 in napi_env__::Unref (this=0x6ec0010) at ../src/js_native_api_v8.h:77
#24 0x00000000010cf741 in v8impl::(anonymous namespace)::ThreadSafeFunction::~ThreadSafeFunction (this=0x6ec1c20, __in_chrg=<optimized out>)
    at ../src/node_api.cc:154
#25 0x00000000010cf7aa in v8impl::(anonymous namespace)::ThreadSafeFunction::~ThreadSafeFunction (this=0x6ec1c20, __in_chrg=<optimized out>)
    at ../src/node_api.cc:155
#26 0x00000000010cfb85 in v8impl::(anonymous namespace)::ThreadSafeFunction::EmptyQueueAndDelete (this=0x6ec1c20) at ../src/node_api.cc:227
#27 0x00000000010d025a in v8impl::(anonymous namespace)::ThreadSafeFunction::Finalize (this=0x6ec1c20) at ../src/node_api.cc:335
#28 0x00000000010d02e4 in operator() (__closure=0x6db4a38, handle=0x6ec1d48) at ../src/node_api.cc:368
#29 0x00000000010d3bb1 in operator() (this=0x0, handle=0x6ec1d48) at ../src/env-inl.h:435
#30 0x00000000010d3bf1 in _FUN () at ../src/env-inl.h:431
#31 0x0000000002221607 in uv__finish_close (handle=0x6ec1d48) at ../deps/uv/src/unix/core.c:303
#32 0x000000000222164d in uv__run_closing_handles (loop=0x5f8c700 <default_loop_struct>) at ../deps/uv/src/unix/core.c:317
#33 0x000000000222185e in uv_run (loop=0x5f8c700 <default_loop_struct>, mode=UV_RUN_ONCE) at ../deps/uv/src/unix/core.c:395
#34 0x00000000010612da in node::Environment::CleanupHandles (this=0x6e6f210) at ../src/env.cc:643
#35 0x000000000106185e in node::Environment::RunCleanup (this=0x6e6f210) at ../src/env.cc:704
#36 0x0000000000feb493 in node::FreeEnvironment (env=0x6e6f210) at ../src/api/environment.cc:371
#37 0x0000000000fe6efc in node::FunctionDeleter<node::Environment, &node::FreeEnvironment>::operator() (this=0x7ffe68204e78,
    pointer=0x6e6f210) at ../src/util.h:636
#38 0x0000000000fe6810 in std::unique_ptr<node::Environment, node::FunctionDeleter<node::Environment, &node::FreeEnvironment> >::~unique_ptr
    (this=0x7ffe68204e78, __in_chrg=<optimized out>)
    at /nix/store/xrb5qsxpvhpm0irr9ykfqpvj8a6sk4xh-gcc-10.2.0/include/c++/10.2.0/bits/unique_ptr.h:361
#39 0x00000000011a09ae in node::NodeMainInstance::Run (this=0x7ffe68204ff0, env_info=0x5f7cd20 <node::env_info>)
    at ../src/node_main_instance.cc:135
#40 0x00000000010c8e57 in node::Start (argc=2, argv=0x7ffe68205228) at ../src/node.cc:1083
#41 0x0000000002bd2b7b in main (argc=2, argv=0x7ffe68205228) at ../src/node_main.cc:127
```

**Comment by pimeys** [comment]
Now seeing why the mutex lock aborts:
```gdb
(gdb) p (int) pthread_mutex_lock(mutex)
$2 = 22
```
And from the libc documentation:
```rust
pub const EINVAL: c_int = 22;
```

**Comment by Brooooooklyn** [comment]
@pimeys published 1.3.4

**Comment by pimeys** [comment]
Thank you @Brooooooklyn it is most definitely fixed now.
